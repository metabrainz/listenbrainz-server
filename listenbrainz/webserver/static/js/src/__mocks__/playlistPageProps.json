{
    "currentUser": {
        "id": 1,
        "name": "iliekcomputers"
    },
    "playlist": {
        "playlist": {
            "creator": "iliekcomputers",
            "identifier": "https://listenbrainz.org/playlist/4245ccd3-4f0d-4276-95d6-2e09d87b5546",
            "date": "2021-01-04T14:43:58.359215+00:00",
            "title": "1980s flashback jams",
            "annotation": "Your lame <i>80s</i> music",
            "extension": {
                "https://musicbrainz.org/doc/jspf#playlist": {
                    "public": true,
                    "last_modified_at": "2021-01-18T16:44:22.837342+00:00"
                }
            },
            "track": [
                {
                    "creator": "Endless Boogie",
                    "extension": {
                        "https://musicbrainz.org/doc/jspf#track": {
                            "added_at": "2020-12-24T01:41:44.821939+00:00",
                            "added_by": "troi-bot",
                            "artist_identifier": [
                                "https://musicbrainz.org/artist/bded3a41-acd7-42ff-9402-cb59e35be8f4"
                            ]
                        }
                    },
                    "identifier": "https://musicbrainz.org/recording/385f4381-0f6d-47a0-bc56-07570374f650",
                    "title": "Vibe Killer"
                }
            ]
        }
    },
    "webSocketsServerUrl": "http://localhost:8082",
    "apiUrl": "http://0.0.0.0",
    "labsApiUrl": "http://0.0.0.0"
}
