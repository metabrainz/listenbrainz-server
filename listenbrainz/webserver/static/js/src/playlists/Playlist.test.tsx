import { enableFetchMocks } from "jest-fetch-mock";
import * as React from "react";
import { shallow } from "enzyme";
import * as timeago from "time-ago";
import PlaylistPage from "./Playlist";
import * as playlistPageProps from "../__mocks__/playlistPageProps.json";

enableFetchMocks();

// Font Awesome generates a random hash ID for each icon everytime.
// Mocking Math.random() fixes this
// https://github.com/FortAwesome/react-fontawesome/issues/194#issuecomment-627235075
jest.spyOn(global.Math, "random").mockImplementation(() => 0);

const {
  apiUrl,
  labsApiUrl,
  currentUser,
  playlist,
  webSocketsServerUrl,
} = playlistPageProps;

const props = {
  apiUrl,
  labsApiUrl,
  playlist: playlist as JSPFObject,
  spotify: {
    access_token: "heyo",
    permission: "streaming" as SpotifyPermission,
  },
  currentUser,
  webSocketsServerUrl,
};

describe("PlaylistPage", () => {
  it("renders correctly", () => {
    // Mock timeago (returns an elapsed time string) otherwise snapshot won't match
    timeago.ago = jest.fn().mockImplementation(() => "1 day ago");
    const wrapper = shallow<PlaylistPage>(<PlaylistPage {...props} />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("does not clear the add-a-track input on blur", async () => {
    const wrapper = shallow<PlaylistPage>(<PlaylistPage {...props} />);
    const instance = wrapper.instance();
    let searchInput = wrapper.find(".search");
    expect(instance.state.searchInputValue).toEqual("");
    // @ts-ignore
    expect(searchInput.props().inputValue).toEqual("");

    searchInput.simulate("focus");
    instance.handleInputChange("mysearch", { action: "input-change" });

    expect(instance.state.searchInputValue).toEqual("mysearch");
    searchInput = wrapper.find(".search");
    // @ts-ignore
    expect(searchInput.props().inputValue).toEqual("mysearch");

    // simulate ReactSelect input blur event
    searchInput.simulate("blur");
    instance.handleInputChange("", { action: "input-blur" });

    searchInput = wrapper.find(".search");
    expect(instance.state.searchInputValue).toEqual("mysearch");
    // @ts-ignore
    expect(searchInput.props().inputValue).toEqual("mysearch");

    // simulate ReactSelect menu close event (blur)
    instance.handleInputChange("", { action: "menu-close" });

    searchInput = wrapper.find(".search");
    expect(instance.state.searchInputValue).toEqual("mysearch");
    // @ts-ignore
    expect(searchInput.props().inputValue).toEqual("mysearch");
  });
});
