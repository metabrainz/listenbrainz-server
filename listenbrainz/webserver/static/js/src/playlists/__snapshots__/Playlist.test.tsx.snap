// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaylistPage renders correctly 1`] = `"<div role=\\"main\\"><div class=\\"AlertContainer-container-0-1-10 AlertContainer-bottom-right-0-1-13\\"><div><div></div></div></div><div class=\\"row\\"><div id=\\"playlist\\" class=\\"col-md-8\\"><div class=\\"playlist-details row\\"><h1 class=\\"title\\"><div>1980s flashback jams<span class=\\"dropdown pull-right\\"><button class=\\"btn btn-info dropdown-toggle\\" type=\\"button\\" id=\\"playlistOptionsDropdown\\" data-toggle=\\"dropdown\\" aria-haspopup=\\"true\\" aria-expanded=\\"true\\"><svg aria-labelledby=\\"svg-inline--fa-title-000000000000\\" data-prefix=\\"fas\\" data-icon=\\"cog\\" class=\\"svg-inline--fa fa-cog fa-w-16 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><title id=\\"svg-inline--fa-title-000000000000\\">Options</title><path fill=\\"currentColor\\" d=\\"M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z\\"></path></svg> Options</button><ul class=\\"dropdown-menu dropdown-menu-right\\" aria-labelledby=\\"playlistOptionsDropdown\\"><li><a role=\\"button\\" href=\\"#\\">Duplicate</a></li><li role=\\"separator\\" class=\\"divider\\"></li><li><a data-toggle=\\"modal\\" data-target=\\"#playlistModal\\" role=\\"button\\" href=\\"#\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"pen\\" class=\\"svg-inline--fa fa-pen fa-w-16 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M290.74 93.24l128.02 128.02-277.99 277.99-114.14 12.6C11.35 513.54-1.56 500.62.14 485.34l12.7-114.22 277.9-277.88zm207.2-19.06l-60.11-60.11c-18.75-18.75-49.16-18.75-67.91 0l-56.55 56.55 128.02 128.02 56.55-56.55c18.75-18.76 18.75-49.16 0-67.91z\\"></path></svg> Edit</a></li><li><a data-toggle=\\"modal\\" data-target=\\"#confirmDeleteModal\\" role=\\"button\\" href=\\"#\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"trash-alt\\" class=\\"svg-inline--fa fa-trash-alt fa-w-14 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 448 512\\"><path fill=\\"currentColor\\" d=\\"M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z\\"></path></svg> Delete</a></li><li role=\\"separator\\" class=\\"divider\\"></li><li><a role=\\"button\\" href=\\"#\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fab\\" data-icon=\\"spotify\\" class=\\"svg-inline--fa fa-spotify fa-w-16 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 496 512\\"><path fill=\\"currentColor\\" d=\\"M248 8C111.1 8 0 119.1 0 256s111.1 248 248 248 248-111.1 248-248S384.9 8 248 8zm100.7 364.9c-4.2 0-6.8-1.3-10.7-3.6-62.4-37.6-135-39.2-206.7-24.5-3.9 1-9 2.6-11.9 2.6-9.7 0-15.8-7.7-15.8-15.8 0-10.3 6.1-15.2 13.6-16.8 81.9-18.1 165.6-16.5 237 26.2 6.1 3.9 9.7 7.4 9.7 16.5s-7.1 15.4-15.2 15.4zm26.9-65.6c-5.2 0-8.7-2.3-12.3-4.2-62.5-37-155.7-51.9-238.6-29.4-4.8 1.3-7.4 2.6-11.9 2.6-10.7 0-19.4-8.7-19.4-19.4s5.2-17.8 15.5-20.7c27.8-7.8 56.2-13.6 97.8-13.6 64.9 0 127.6 16.1 177 45.5 8.1 4.8 11.3 11 11.3 19.7-.1 10.8-8.5 19.5-19.4 19.5zm31-76.2c-5.2 0-8.4-1.3-12.9-3.9-71.2-42.5-198.5-52.7-280.9-29.7-3.6 1-8.1 2.6-12.9 2.6-13.2 0-23.3-10.3-23.3-23.6 0-13.6 8.4-21.3 17.4-23.9 35.2-10.3 74.6-15.2 117.5-15.2 73 0 149.5 15.2 205.4 47.8 7.8 4.5 12.9 10.7 12.9 22.6 0 13.6-11 23.3-23.2 23.3z\\"></path></svg> Export to Spotify</a></li></ul></span></div><small>Public playlist by <a href=\\"/user/iliekcomputers/playlists\\">iliekcomputers</a></small></h1><div class=\\"info\\"><div>1 tracks</div><div>Created: 1/4/2021, 2:43:58 PM</div><div>Last modified: 1/18/2021, 4:44:22 PM</div></div><div>Your lame <i>80s</i> music</div><hr/></div><div id=\\"listens row\\"><div><div style=\\"background:#FFFFFF;border:1px solid #EEEEEE;box-sizing:border-box;box-shadow:0 1px 1px rgba(192,192,192,0.1), 0 2px 2px rgba(192,192,192,0.15), 0 4px 4px rgba(192,192,192,0.20);border-radius:8px;height:100%\\" class=\\"playlist-item-card row  \\" data-recording-mbid=\\"385f4381-0f6d-47a0-bc56-07570374f650\\"><svg aria-labelledby=\\"svg-inline--fa-title-000000000000\\" data-prefix=\\"fas\\" data-icon=\\"grip-lines\\" class=\\"svg-inline--fa fa-grip-lines fa-w-16 drag-handle text-muted\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><title id=\\"svg-inline--fa-title-000000000000\\">Drag to reorder</title><path fill=\\"currentColor\\" d=\\"M496 288H16c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h480c8.8 0 16-7.2 16-16v-32c0-8.8-7.2-16-16-16zm0-128H16c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h480c8.8 0 16-7.2 16-16v-32c0-8.8-7.2-16-16-16z\\"></path></svg><div class=\\"track-details\\"><div title=\\"Vibe Killer\\"><a href=\\"https://musicbrainz.org/recording/385f4381-0f6d-47a0-bc56-07570374f650\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">Vibe Killer</a></div><small class=\\"text-muted\\" title=\\"Endless Boogie\\"><a href=\\"https://musicbrainz.org/artist/bded3a41-acd7-42ff-9402-cb59e35be8f4\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">Endless Boogie</a></small></div><div class=\\"addition-details\\">added <span class=\\"listen-time\\" title=\\"12/24/2020, 1:41:44 AM\\">1 day ago</span><div>by troi-bot</div></div><span class=\\"dropdown\\"><button class=\\"btn btn-link btn-sm dropdown-toggle\\" type=\\"button\\" id=\\"listenControlsDropdown\\" data-toggle=\\"dropdown\\" aria-haspopup=\\"true\\" aria-expanded=\\"true\\"><svg aria-labelledby=\\"svg-inline--fa-title-000000000000\\" data-prefix=\\"fas\\" data-icon=\\"ellipsis-v\\" class=\\"svg-inline--fa fa-ellipsis-v fa-w-6 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 192 512\\"><title id=\\"svg-inline--fa-title-000000000000\\">More options</title><path fill=\\"currentColor\\" d=\\"M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z\\"></path></svg></button><ul class=\\"dropdown-menu dropdown-menu-right\\" aria-labelledby=\\"listenControlsDropdown\\"><li><a href=\\"//musicbrainz.org/recording/385f4381-0f6d-47a0-bc56-07570374f650\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">Open in MusicBrainz</a></li><li><button type=\\"button\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"trash-alt\\" class=\\"svg-inline--fa fa-trash-alt fa-w-14 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 448 512\\"><path fill=\\"currentColor\\" d=\\"M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z\\"></path></svg> Remove from playlist</button></li></ul></span></div></div><div style=\\"background:#FFFFFF;border:1px solid #EEEEEE;box-sizing:border-box;box-shadow:0 1px 1px rgba(192,192,192,0.1), 0 2px 2px rgba(192,192,192,0.15), 0 4px 4px rgba(192,192,192,0.20);border-radius:8px;height:100%\\" class=\\"playlist-item-card row\\" id=\\"add-track\\"><span><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"plus-circle\\" class=\\"svg-inline--fa fa-plus-circle fa-w-16 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z\\"></path></svg>  Add a track</span><div class=\\"search css-2b097c-container\\"><div class=\\" css-yk16xz-control\\"><div class=\\" css-g1d714-ValueContainer\\"><div class=\\" css-1wa3eu0-placeholder\\">Artist followed by track name</div><div class=\\"css-b8ldur-Input\\"><div class=\\"\\" style=\\"display:inline-block\\"><input type=\\"text\\" autoCapitalize=\\"none\\" autoComplete=\\"off\\" autoCorrect=\\"off\\" id=\\"react-select-2-input\\" spellcheck=\\"false\\" tabindex=\\"0\\" value=\\"\\" aria-autocomplete=\\"list\\" style=\\"box-sizing:content-box;width:1px;label:input;background:0;border:0;font-size:inherit;opacity:1;outline:0;padding:0;color:inherit\\"/><div style=\\"position:absolute;top:0;left:0;visibility:hidden;height:0;overflow:scroll;white-space:pre\\"></div></div></div></div><div class=\\" css-1hb7zxy-IndicatorsContainer\\"><span class=\\" css-1okebmr-indicatorSeparator\\"></span><div aria-hidden=\\"true\\" class=\\" css-tlfecz-indicatorContainer\\"><svg height=\\"20\\" width=\\"20\\" viewBox=\\"0 0 20 20\\" aria-hidden=\\"true\\" focusable=\\"false\\" class=\\"css-6q0nyr-Svg\\"><path d=\\"M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z\\"></path></svg></div></div></div></div></div></div><div class=\\"modal fade\\" id=\\"playlistModal\\" tabindex=\\"-1\\" role=\\"dialog\\" aria-labelledby=\\"playlistModalLabel\\"><div class=\\"modal-dialog\\" role=\\"document\\"><form class=\\"modal-content\\"><div class=\\"modal-header\\"><button type=\\"button\\" class=\\"close\\" data-dismiss=\\"modal\\" aria-label=\\"Close\\"><span aria-hidden=\\"true\\">×</span></button><h4 class=\\"modal-title\\" id=\\"playlistModalLabel\\">Edit playlist</h4></div><div class=\\"modal-body\\"><div class=\\"form-group\\"><label for=\\"playlistName\\">Name</label><input type=\\"text\\" class=\\"form-control\\" id=\\"playlistName\\" placeholder=\\"Name\\" value=\\"1980s flashback jams\\" name=\\"name\\"/></div><div class=\\"form-group\\"><label for=\\"playlistdescription\\">Description</label><textarea class=\\"form-control\\" id=\\"playlistdescription\\" placeholder=\\"Description\\" name=\\"description\\">Your lame &lt;i&gt;80s&lt;/i&gt; music</textarea></div><div class=\\"checkbox\\"><label><input type=\\"checkbox\\" id=\\"isPublic\\" checked=\\"\\" name=\\"isPublic\\"/> Make playlist public</label></div><div class=\\"form-group\\"><label for=\\"playlistcollaborators\\">Collaborators</label><table id=\\"playlistcollaborators\\" class=\\"table table-condensed table-striped listens-table\\"><tbody></tbody></table><div class=\\"input-group input-group-flex\\"><span class=\\"input-group-addon\\">Add collaborator</span><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Username…\\" name=\\"newCollaborator\\" value=\\"\\"/><span class=\\"input-group-btn\\"><button class=\\"btn\\" type=\\"submit\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"plus-circle\\" class=\\"svg-inline--fa fa-plus-circle fa-w-16 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z\\"></path></svg> Add</button></span></div></div></div><div class=\\"modal-footer\\"><button type=\\"button\\" class=\\"btn btn-default\\" data-dismiss=\\"modal\\">Cancel</button><button type=\\"submit\\" class=\\"btn btn-primary\\" data-dismiss=\\"modal\\">Save</button></div></form></div></div><div class=\\"modal fade\\" id=\\"confirmDeleteModal\\" tabindex=\\"-1\\" role=\\"dialog\\" aria-labelledby=\\"confirmDeleteModalLabel\\"><div class=\\"modal-dialog modal-sm\\" role=\\"document\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><button type=\\"button\\" class=\\"close\\" data-dismiss=\\"modal\\" aria-label=\\"Close\\"><span aria-hidden=\\"true\\">×</span></button><h4 class=\\"modal-title\\" id=\\"confirmDeleteModalLabel\\">Delete playlist</h4></div><div class=\\"modal-body\\">You are about to delete playlist <i>1980s flashback jams</i>.<br/>This action cannot be undone.</div><div class=\\"modal-footer\\"><button type=\\"button\\" class=\\"btn btn-default\\" data-dismiss=\\"modal\\">Cancel</button><button type=\\"button\\" class=\\"btn btn-danger\\" data-dismiss=\\"modal\\">Confirm</button></div></div></div></div></div><div class=\\"col-md-4\\" style=\\"position:sticky;top:20px\\"><div><div id=\\"music-player\\" aria-label=\\"Playback control\\"><div class=\\"content\\"><div class=\\"youtube hidden\\"><div class=\\"\\"><div></div></div></div><div class=\\"soundcloud hidden\\"><iframe id=\\"soundcloud-iframe\\" title=\\"Soundcloud player\\" width=\\"100%\\" height=\\"420px\\" scrolling=\\"no\\" frameBorder=\\"no\\" allow=\\"autoplay\\" src=\\"https://w.soundcloud.com/player/?url=\\"></iframe></div><div class=\\"no-album-art\\">No album art</div></div><div class=\\"info showControls\\"><div class=\\"currently-playing\\"><div class=\\"progress\\" aria-label=\\"Audio Progress Control\\" role=\\"progressbar\\" aria-valuemin=\\"0\\" aria-valuemax=\\"100\\" aria-valuenow=\\"NaN\\" tabindex=\\"0\\"><div class=\\"progress-bar\\" style=\\"width:NaN%\\"></div></div></div><div class=\\"controls\\"><div class=\\"left btn btn-xs\\" title=\\"Always show controls\\" role=\\"button\\" tabindex=\\"0\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"eye-slash\\" class=\\"svg-inline--fa fa-eye-slash fa-w-20 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 640 512\\"><path fill=\\"currentColor\\" d=\\"M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z\\"></path></svg></div><div class=\\"previous btn btn-xs\\" title=\\"Previous\\" role=\\"button\\" tabindex=\\"0\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"fast-backward\\" class=\\"svg-inline--fa fa-fast-backward fa-w-16 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M0 436V76c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v151.9L235.5 71.4C256.1 54.3 288 68.6 288 96v131.9L459.5 71.4C480.1 54.3 512 68.6 512 96v320c0 27.4-31.9 41.7-52.5 24.6L288 285.3V416c0 27.4-31.9 41.7-52.5 24.6L64 285.3V436c0 6.6-5.4 12-12 12H12c-6.6 0-12-5.4-12-12z\\"></path></svg></div><div class=\\"play btn\\" title=\\"Play\\" role=\\"button\\" tabindex=\\"0\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"play-circle\\" class=\\"svg-inline--fa fa-play-circle fa-w-16 fa-2x \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm115.7 272l-176 101c-15.8 8.8-35.7-2.5-35.7-21V152c0-18.4 19.8-29.8 35.7-21l176 107c16.4 9.2 16.4 32.9 0 42z\\"></path></svg></div><div class=\\"next btn btn-xs\\" title=\\"Next\\" role=\\"button\\" tabindex=\\"0\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"fast-forward\\" class=\\"svg-inline--fa fa-fast-forward fa-w-16 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M512 76v360c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12V284.1L276.5 440.6c-20.6 17.2-52.5 2.8-52.5-24.6V284.1L52.5 440.6C31.9 457.8 0 443.4 0 416V96c0-27.4 31.9-41.7 52.5-24.6L224 226.8V96c0-27.4 31.9-41.7 52.5-24.6L448 226.8V76c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12z\\"></path></svg></div><div class=\\"right btn btn-xs\\" title=\\"Play up\\" role=\\"button\\" tabindex=\\"0\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"sort-amount-down\\" class=\\"svg-inline--fa fa-sort-amount-down fa-w-16 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M304 416h-64a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-128-64h-48V48a16 16 0 0 0-16-16H80a16 16 0 0 0-16 16v304H16c-14.19 0-21.37 17.24-11.29 27.31l80 96a16 16 0 0 0 22.62 0l80-96C197.35 369.26 190.22 352 176 352zm256-192H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h192a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-64 128H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h128a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zM496 32H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h256a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z\\"></path></svg></div></div></div></div></div></div></div></div>"`;
